### DISCUSSION TITLE
How to set the time/time zone correctly?

### DISCUSSION THREAD

**Discussion by home409ca** [wlcrs/huawei_solar#501] [General]
The inverter started at 07:09

The "Inverter Startup time" entity incorrectly shows 08:09

Poland is currently in UTC+2 but soon it will be in UTC+1 for half a year)

What to do to make the entity value correct?
Thanks

![image](https://github.com/wlcrs/huawei_solar/assets/114701129/e91e4eef-ac85-4449-9cf7-88964c79e359)

**Comment by Roving-Ronin** [comment]
@home409ca 
I don't think this is anything to do with the WLCRS integration. 
Looking through my own logs this appears to be an issue with HA, and is also affecting my own automations and sensors, such as I use to update the tariff rates for utility sensors and energy_meter sensors that also track electricity / gas usage and costs. 
This is even though HA is configured for +11 timezone (Sydney, Australia) and it shows the correct time, all the automations/triggers/logs are show and working an hour later than expected / set.

Currently scratching head and trying to ascertain is having some 'Is it DST' = No (0) / Yes (1) sensor would be beneficial, to then add in conditions to say is DST = Yes(1) then run 1 hour earlier.

**Comment by home409ca** [comment]
Let the author speak. Or maybe someone else has a solution?

**Comment by wlcrs** [comment]
The integration offsets the retrieved timestamps against the timezone it retrieves from the inverter: https://github.com/wlcrs/huawei-solar-lib/blob/master/src/huawei_solar/registers.py#L239

![image](https://github.com/wlcrs/huawei_solar/assets/2150060/ca48af7f-7b51-4935-9c24-7613ae34e154)

In my installation this also results in results that are skewed by one hour:

![image](https://github.com/wlcrs/huawei_solar/assets/2150060/0822e9cc-248a-4d10-b2bf-d42b8c0f06f6)

To be honest, I haven't bothered trying to correct this, as with the information of only my own inverter it's very hard to understand what is going on. If anyone figures out the correct formula, please let me know and I'll fix the code accordingly.

**Reply by home409ca** [reply]
Will this work? I wouldn't want to spoil it.
I'm not a programmer, I asked GPT chat.

```
from datetime import datetime, timezone, timedelta
.
.
.
return datetime.fromtimestamp(value - 60 * inverter.time_zone, timezone.utc) - timedelta(hours=1)
```

**Comment by home409ca** [comment]
I looked for the registers.py file in custom_components to check it myself, but I didn't find it.

Where can I add "- timedelta(hours=1)"?

Or simply add "time correction" to the add-on configuration, because I see that it's not only me who has this problem.
